src/core/util/param.pl

```diff
@@ -0,0 +1,122 @@
+/*
+ * This is a collection of utilities for finding, checking, and converting
+ * values from JSON and search (a.k.a. query string parameters).
+ *
+ * All JSON and search values should involve one of these predicates so that we
+ * can have consistent and safe input validation along with minimal redundancy.
+ */
+:- module(param, [
+              param_check_json/4,
+              param_check_search/4,
+
+              param_value_search/4,
+              param_value_json/4,
+              param_value_search_or_json/5,
+
+              param_value_search_author/2,
+              param_value_search_message/2,
+              param_value_search_graph_type/2
+          ]).
+
+/*
+ * Check a JSON dict parameter value. If it's wrong, throw an error. Otherwise,
+ * convert it if necessary.
+ */
+param_check_json(boolean, _, false, false).
+param_check_json(boolean, _, true, true).
+param_check_json(graph, _, "schema", schema).
+param_check_json(graph, _, "instance", instance).
+param_check_json(object, _, Value_In, Value_Out) :-
+    is_dict(Value_In),
+    !,
+    Value_Out = Value_In.
+param_check_json(integer, _, Value_In, Value_Out) :-
+    input_to_integer(Value_In, Value_Out),
+    !,
+    Value_Out = Value_In.
+param_check_json(nonnegative_integer, _, Value_In, Value_Out) :-
+    input_to_integer(Value_In, Value_Integer),
+    Value_Integer >= 0,
+    !,
+    Value_Out = Value_Integer.
+param_check_json(string, _, Value_In, Value_Out) :-
+    string(Value_In),
+    Value_In \= "",
+    !,
+    Value_Out = Value_In.
+param_check_json(atom_or_string, Key, Value_In, Value_Out) :-
+    param_check_json(string, Key, Value_In, Value_String),
+    !,
+    atom_string(Value_Out, Value_String).
+param_check_json(Type, Key, Value_In, _) :-
```

Comment:
No serious thing especially in this case but it is slightly better stylistically and in performance to make an inner predicate which is semi-det and first argument indexed on an atom, and then have an outer predicate which is det+exception which judges if the inner predicate fails or not. 

This way no cuts will be required.

Reply:
e.g.

```prolog
param_check_json(Type, Key, Value_In, Value_Out) :- 
  (    param_check_json_(Type, Key, Value_In, Value_Out)
  ->  true
  ;    throw(error(bad_parameter_type), _)
  ) .

param_check_json_(boolean, _, false, false).
param_check_json_(boolean, _, true, true).
param_check_json_(graph, _, "schema", schema).
param_check_json_(graph, _, "instance", instance).
param_check_json_(object, _, Value_In, Value_Out) :-
    is_dict(Value_In),
    Value_Out = Value_In.
...
```

Reply:
Good style. Done!